openmls/src/group/mls_group/past_secrets.rs

```diff
@@ -0,0 +1,60 @@
+use std::collections::VecDeque;
+
+use crate::schedule::message_secrets::MessageSecrets;
+
+use super::*;
+
+// Internal helper struct
+#[derive(Debug, Serialize, Deserialize)]
+struct EpochTree {
+    epoch: u64,
+    message_secrets: MessageSecrets,
+}
+
+/// Can store message secrets for up to `max_epochs`. The trees are added with [`self::add()`] and can be queried
+/// with [`Self::get_epoch()`].
+#[derive(Debug, Serialize, Deserialize)]
+pub struct MessageSecretsStore {
+    max_epochs: usize,
+    epoch_trees: VecDeque<EpochTree>,
+}
+
+impl MessageSecretsStore {
+    /// Create a new store that can hold up to `max_epochs` message secrets.
+    /// If `max_epochs` is 0, no secret trees will be stored.
+    pub fn new(max_epochs: usize) -> Self {
+        Self {
+            max_epochs,
+            epoch_trees: VecDeque::new(),
+        }
+    }
+
+    /// Add a secret tree for a given epoch `group_epoch`.
+    pub fn add(&mut self, group_epoch: GroupEpoch, message_secrets: MessageSecrets) {
+        // Don't store the tree if it's not intended
+        if self.max_epochs == 0 {
+            return;
+        }
+        let GroupEpoch(epoch) = group_epoch;
+        let epoch_tree = EpochTree {
+            epoch,
+            message_secrets,
+        };
+        while self.epoch_trees.len() >= self.max_epochs {
+            self.epoch_trees.pop_front();
```

Comment:
Should we be able to clear the store without adding? Not sure if it's needed. But I feel like there should be a way to delete secrets explicitely.

Reply:
So far you can only drop the whole store if you want to delete secrets. I didn't add a more fine-grained method because I didn't want to write tests for something we don't need right now. Adding it with a test is easy though if we ever need it.